subsys/settings/src/settings_zms.c

```diff
@@ -96,117 +133,262 @@ static int settings_zms_unlink_ll_node(struct settings_zms *cf, uint32_t name_ha
-	if (rc < 0) {
-		return rc;
+#endif /* CONFIG_SETTINGS_ZMS_NO_LL_DELETE */
+
+	return rc;
+}
+
+#ifdef CONFIG_SETTINGS_ZMS_LOAD_SUBTREE_PATH
+/* Loads first the key which is defined by the name found in "subtree" root.
+ * If the key is not found or further keys under the same subtree are needed
+ * by the caller, returns 0.
+ */
+static int settings_zms_load_subtree(struct settings_store *cs, const struct settings_load_arg *arg)
+{
+	struct settings_zms *cf = CONTAINER_OF(cs, struct settings_zms, cf_store);
+	struct settings_zms_read_fn_arg read_fn_arg;
+	char name[SETTINGS_FULL_NAME_LEN];
+	ssize_t rc1;
+	ssize_t rc2;
+	uint32_t name_hash;
+
+	name_hash = sys_hash32(arg->subtree, strnlen(arg->subtree, SETTINGS_FULL_NAME_LEN)) &
+		    ZMS_HASH_MASK;
+	for (int i = 0; i <= cf->hash_collision_num; i++) {
+		name_hash = ZMS_UPDATE_COLLISION_NUM(name_hash, i);
+		/* Get the name entry from ZMS */
+		rc1 = zms_read(&cf->cf_zms, ZMS_NAME_ID_FROM_HASH(name_hash), &name,
+			       sizeof(name) - 1);
+		/* get the length of data and verify that it exists */
+		rc2 = zms_get_data_length(&cf->cf_zms, ZMS_DATA_ID_FROM_HASH(name_hash));
+		if ((rc1 <= 0) || (rc2 <= 0)) {
+			/* Name or data doesn't exist */
+			continue;
+		}
+		/* Found a name, this might not include a trailing \0 */
+		name[rc1] = '\0';
+		if (strcmp(arg->subtree, name)) {
+			/* Names are not equal let's continue to the next collision hash
+			 * if it exists.
+			 */
+			continue;
+		}
+		/* At this steps the names are equal, let's set the handler */
+		read_fn_arg.fs = &cf->cf_zms;
+		read_fn_arg.id = ZMS_DATA_ID_FROM_HASH(name_hash);
+
+		/* We should return here as there is no need to look for the next
+		 * hash collision
+		 */
+		return settings_call_set_handler(arg->subtree, rc2, settings_zms_read_fn,
+						 &read_fn_arg, arg);
 	}
 
-	/* Now delete the current linked list element */
-	rc = zms_delete(&cf->cf_zms, name_hash | 1);
-	if (rc < 0) {
-		return rc;
+	return 0;
+}
+#endif /* CONFIG_SETTINGS_ZMS_LOAD_SUBTREE_PATH */
+
+/* Search for the name_hash that corresponds to name.
+ * If no hash that corresponds to name is found in the persistent storage,
+ * returns 0.
+ */
+static uint32_t settings_zms_find_hash_from_name(struct settings_zms *cf, const char *name)
+{
+	uint32_t name_hash = 0;
+	int rc = 0;
+	char r_name[SETTINGS_FULL_NAME_LEN];
+
+	name_hash = sys_hash32(name, strnlen(name, SETTINGS_FULL_NAME_LEN)) & ZMS_HASH_MASK;
+	for (int i = 0; i <= cf->hash_collision_num; i++) {
+		name_hash = ZMS_UPDATE_COLLISION_NUM(name_hash, i);
+		/* Get the name entry from ZMS */
+		rc = zms_read(&cf->cf_zms, ZMS_NAME_ID_FROM_HASH(name_hash), r_name,
+			      sizeof(r_name) - 1);
+		if (rc <= 0) {
+			/* Name doesn't exist */
+			continue;
+		}
+		/* Found a name, this might not include a trailing \0 */
+		r_name[rc] = '\0';
+		if (strcmp(name, r_name)) {
+			/* Names are not equal let's continue to the next collision hash
+			 * if it exists.
+			 */
+			continue;
+		}
```

Comment:
You have already done `strnlen` of `name` in line 223, so you may as well cache it and split this part to length comparisoin + `memcmp`, should be faster as `memcmp` does not have to search for terminating null.